### PR TITLE
Migrate virt_spec to Java

### PR DESCRIPTION
**Pull request by Januson**

### CODE REVIEW

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+            .listPoolsByProduct(owner.getId(), product.getId());
+        assertThat(poolsAfterBind).hasSize(3);
+        assertThat(findNormalPools(poolsAfterBind))
+            .filteredOn(pool -> !normalPool.getId().equals(pool.getId()))
+            .hasSize(1)
+            .map(PoolDTO::getQuantity)
+            .containsOnly(3L);
+
+        hostClient.entitlements()
+            .updateEntitlement(entitlement.get(0).get("id").asText(), toUpdate(entitlement));
```
Comment (nikosmoum) [inline spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java:621]: shouldn't this be reducing the quantity from 3 to 2? `toUpdate` it seems to just pass the original entitlement's quantity (3)

**Review by nikosmoum — CHANGES_REQUESTED** [review]
A few minor issues but otherwise good

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+        otherHostClient.consumers().bindProduct(otherHost.getUuid(), datacenterProduct2.getId());
+
+        linkHostToGuests(hostClient, host, OTHER_GUEST_UUID);
+        assertThatEntitlementsOf(otherGuestClient, otherGuest).isEmpty();
+
+        otherGuestClient.consumers().autoBind(otherGuest.getUuid());
+        assertThatEntitlementsOf(otherGuestClient, otherGuest).hasSize(1);
+
+        assertThatEntitlementsOf(otherGuestClient, otherGuest)
+            .hasSize(1)
```
[the remark this answers is not in the document]
Comment (Januson) [reply spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java]: Done. I forgot to remove it after refactoring it below. There should be only one entitlement. There is one that covers both products. The second entitlement in the original test was caused by global setup in virt_helper.

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+        guestClient.consumers().bindPool(guest.getUuid(), guestPool.getId(), 1);
+        assertThatEntitlementsOf(guestClient, guest).hasSize(1);
+
+        hostClient.guestIds().deleteGuest(host.getUuid(), GUEST_UUID, false);
+
+        assertThatEntitlementsOf(guestClient, guest).hasSize(1);
+    }
+
+    @Test
+    public void shouldLoseEntitlementWhenGuestStopsAndIsRestartedElsewhere() {
```
[the remark this answers is not in the document]
Comment (Januson) [reply spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java:355]: Done. I missed a piece during migration.

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+                    Map.entry("virtWhoType", "libvirt")
+                )
+            ))
+            .collect(Collectors.toList());
+        linkHostToGuests(hostClient, host, guestIds);
+    }
+
+    private void linkHostToGuests(ApiClient hostClient, ConsumerDTO host, List<GuestIdDTO> guestIds) {
+        hostClient.consumers().updateConsumer(host.getUuid(), host.guestIds(guestIds));
+//        hostClient.guestIds().updateGuests(host.getUuid(), guestIds);
```
Comment (nikosmoum) [inline spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java:749]: accidentally left over?

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+        linkHostToGuests(hostClient, host, GUEST_UUID);
+        ProductDTO virtLimitProduct = createVirtLimitProduct("2");
+        PoolDTO virtLimitPool = admin.owners().
+            createPool(owner.getKey(), Pools.randomUpstream(virtLimitProduct));
+        hostClient.consumers().bindPool(host.getUuid(), virtLimitPool.getId(), 1);
+
+        PoolDTO guestPool = findGuestPool(guestClient, guest);
+        guestClient.consumers().bindPool(guest.getUuid(), guestPool.getId(), 1);
+
+        linkHostToGuests(hostClient, host, GUEST_UUID);
```
Comment (nikosmoum) [inline spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java:367]: this should probably be `unlinkHostsGuests`

**Review by nikosmoum — CHANGES_REQUESTED** [review]

File: spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java (hunk 4 above)
Comment (Januson) [reply spec-tests/src/test/java/org/candlepin/spec/VirtSpecTest.java]: Done.

**Pull request merged**